sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -1546,16 +1550,32 @@ private[lf] object Speedy {
               case V.ValueVariant(_, constructor, value) =>
                 val info =
                   assertRight(
-                    compiledPackages.pkgInterface.lookupVariantConstructor(tyCon, constructor)
+                    compiledPackages.pkgInterface.lookupVariantConstructor(tyCon, constructor),
+                    withLookupError = { case LookupError.NotFound(_, _) =>
+                      throw SErrorDamlException(
+                        IError.Upgrade(
+                          IError.Upgrade.DowngradeFailed(ty, value)
+                        )
+                      )
+                    },
                   )
                 val valType = info.concreteType(argTypes)
                 SValue.SVariant(tyCon, constructor, info.rank, go(valType, value))
+
               case V.ValueEnum(_, constructor) =>
                 val rank =
                   assertRight(
-                    compiledPackages.pkgInterface.lookupEnumConstructor(tyCon, constructor)
+                    compiledPackages.pkgInterface.lookupEnumConstructor(tyCon, constructor),
+                    withLookupError = { case LookupError.NotFound(_, _) =>
+                      throw SErrorDamlException(
+                        IError.Upgrade(
+                          IError.Upgrade.DowngradeFailed(ty, value)
+                        )
+                      )
+                    },
                   )
```

Comment:
ideally you should continue to throw the lookup error if the one of the following is not found: package, module, definition, variant  and report an upgrade error iff the constructor is missing

```suggestion
                val rank =
                  assertRight(
                    compiledPackages.pkgInterface.lookupEnumConstructor(tyCon, constructor),
                    withLookupError = {
                      case LookupError.NotFound(_, _) =>
                        throw SErrorDamlException(
                          IError.Upgrade(
                            IError.Upgrade.DowngradeFailed(ty, value)
                          )
                        )
                    }
                  )
```
or  without catching
```suggestion
                val variantInfo = assertRight(compiledPackages.pkgInterface.lookupDataVariant(tyCon)
                val rank = variantInfo.dataVariant.constructorInfo.get(constructor) match {
                  case Some(r) => r
                  case None => throw SErrorDamlException(
                    IError.Upgrade(
                      IError.Upgrade.DowngradeFailed(ty, value)
                    )
                  )
                }
```

Reply:
> deally you should continue to throw the lookup error if the one of the following is not found: package, module, definition, variant and report an upgrade error iff the constructor is missing

for `lookupEnumConstructor` and `lookupVariantConstructor`, all instances of `LookupError.NotFound` are due to one of these lookup failures. 

So the code as it stands should have these required properties. Are you suggesting I should make these requirements more explicit in the code by having a specific set of variant/enum lookup functions for `importValue`?

> or without catching

`assertRight` is matching on `Either` values and so isn't really throwing or catching anything? As noted above, perhaps you're after a more direct way of encoding the above requirements?

Reply:
`compiledPackages.pkgInterface.lookupEnumConstructor(tyCon, constructor)` can return 4 type of `LookupError.NotFound`:
- `LookupError.NotFound(Reference.Module(pkgId, modName), context)`
- `Left(LookupError.NotFound(Reference.Definition(name), context))
- `Left(NotFound(Reference.DataType(name), context))` 
- `Left(NotFound(Reference.DataEnum(tyCon), context))`
- `Left(NotFound(DataEnumConstructor(tyCon, constructor), context))`